File: sdk/serviceability/typescript/serviceability/tests/pda.test.ts
import { describe, expect, test } from "bun:test";
import { PublicKey } from "@solana/web3.js";
import {
  deriveGlobalStatePda,
  deriveGlobalConfigPda,
  deriveProgramConfigPda,
} from "../pda.js";

const PROGRAM_ID = new PublicKey(
  "ser2VaTMAcYTaauMrTSfSrxBaUDq7BLNs2xfUugTAGv",
);

describe("PDA derivation", () => {
  test("global state PDA is deterministic", () => {
    const [addr1, bump1] = deriveGlobalStatePda(PROGRAM_ID);
    const [addr2, bump2] = deriveGlobalStatePda(PROGRAM_ID);
    expect(addr1.equals(addr2)).toBe(true);
    expect(bump1).toBe(bump2);
  });

  test("global config PDA", () => {
    const [addr] = deriveGlobalConfigPda(PROGRAM_ID);
    expect(addr.equals(PublicKey.default)).toBe(false);
  });

  test("program config PDA", () => {
    const [addr] = deriveProgramConfigPda(PROGRAM_ID);
    expect(addr.equals(PublicKey.default)).toBe(false);
  });

  test("PDAs are different", () => {
    const [gs] = deriveGlobalStatePda(PROGRAM_ID);
    const [gc] = deriveGlobalConfigPda(PROGRAM_ID);
    const [pc] = deriveProgramConfigPda(PROGRAM_ID);
    expect(gs.equals(gc)).toBe(false);
    expect(gs.equals(pc)).toBe(false);
    expect(gc.equals(pc)).toBe(false);
  });
});

File: sdk/serviceability/go/pda_test.go
package serviceability

import (
	"testing"

	"github.com/gagliardetto/solana-go"
)

var testProgramID = solana.MustPublicKeyFromBase58("ser2VaTMAcYTaauMrTSfSrxBaUDq7BLNs2xfUugTAGv")

func TestDeriveGlobalStatePDA(t *testing.T) {
	addr, bump, err := DeriveGlobalStatePDA(testProgramID)
	if err != nil {
		t.Fatalf("DeriveGlobalStatePDA: %v", err)
	}
	if addr.IsZero() {
		t.Error("derived zero address")
	}
	if bump == 0 {
		t.Log("bump is 0 (valid but unusual)")
	}

	addr2, bump2, err := DeriveGlobalStatePDA(testProgramID)
	if err != nil {
		t.Fatalf("DeriveGlobalStatePDA (2nd): %v", err)
	}
	if addr != addr2 || bump != bump2 {
		t.Error("PDA derivation not deterministic")
	}
}

func TestDeriveGlobalConfigPDA(t *testing.T) {
	addr, _, err := DeriveGlobalConfigPDA(testProgramID)
	if err != nil {
		t.Fatalf("DeriveGlobalConfigPDA: %v", err)
	}
	if addr.IsZero() {
		t.Error("derived zero address")
	}
}

func TestDeriveProgramConfigPDA(t *testing.T) {
	addr, _, err := DeriveProgramConfigPDA(testProgramID)
	if err != nil {
		t.Fatalf("DeriveProgramConfigPDA: %v", err)
	}
	if addr.IsZero() {
		t.Error("derived zero address")
	}
}

func TestPDAsAreDifferent(t *testing.T) {
	gs, _, _ := DeriveGlobalStatePDA(testProgramID)
	gc, _, _ := DeriveGlobalConfigPDA(testProgramID)
	pc, _, _ := DeriveProgramConfigPDA(testProgramID)
	if gs == gc || gs == pc || gc == pc {
		t.Error("different PDA types produced same address")
	}
}
